tools/mirror/src/lib.rs

```diff
@@ -0,0 +1,693 @@
+        source_block_hash: &CryptoHash,
+    ) -> anyhow::Result<Result<Nonce, MapNonceError>> {
+        let db_key = target_public.try_to_vec().unwrap();
+        let mut m =
+            // TODO: cache this?
+            match self.db.get_cf(self.db.cf_handle(DBCol::Nonces.name()).unwrap(), &db_key)? {
+                Some(v) => NonceDiff::try_from_slice(&v).unwrap(),
+                // If it's not stored in the database, it's an access key that was present in the genesis
+                // recods, so we don't need to do anything to the nonce.
+                None => return Ok(Ok(nonce)),
+            };
+        if m.known() {
+            return Ok(m.map(nonce));
+        }
+
+        let source_nonce = fetch_access_key_nonce(
+            &self.source_view_client,
+            signer_id,
+            source_public,
+            Some(source_block_hash),
+        )
+        .await?;
+
+        let target_nonce =
+            fetch_access_key_nonce(&self.target_view_client, signer_id, target_public, None)
+                .await?
+                // add one because we need to start from a nonce one greater than the current one
+                .map(|n| n + 1);
+
+        let mut rewrite = false;
+        if m.source_start.is_none() && source_nonce.is_some() {
+            rewrite = true;
+            m.source_start = source_nonce;
+        }
+        if m.target_start.is_none() && target_nonce.is_some() {
+            rewrite = true;
+            m.target_start = target_nonce;
+        }
+        if rewrite {
+            tracing::debug!(target: "mirror", "storing {:?} in DB for ({:?}, {:?})", &m, signer_id, target_public);
+            self.db.put_cf(
+                self.db.cf_handle(DBCol::Nonces.name()).unwrap(),
+                &db_key,
+                &m.try_to_vec().unwrap(),
+            )?;
+        }
+        Ok(m.map(nonce))
+    }
+
+    fn map_action(&self, a: &ActionView) -> anyhow::Result<Option<Action>> {
+        // this try_from() won't fail since the ActionView was constructed from the Action
+        let action = Action::try_from(a.clone()).unwrap();
+
+        match &action {
+            Action::AddKey(add_key) => {
+                let replacement =
+                    crate::key_mapping::map_key(&add_key.public_key, self.secret.as_ref());
+                let public_key = replacement.public_key();
+                self.db.put_cf(
+                    self.db.cf_handle(DBCol::Nonces.name()).unwrap(),
+                    &public_key.try_to_vec().unwrap(),
+                    &NonceDiff::default().try_to_vec().unwrap(),
+                )?;
+
+                Ok(Some(Action::AddKey(AddKeyAction {
+                    public_key,
+                    access_key: add_key.access_key.clone(),
+                })))
+            }
+            Action::DeleteKey(delete_key) => {
+                let replacement =
+                    crate::key_mapping::map_key(&delete_key.public_key, self.secret.as_ref());
```

Comment:
I don't understand how this is supposed to work.
If an account has 10 keys, and 1 gets deleted. Will this result in deleting shardnet's secret key, even though the account has 9 other keys? Each of the remaining keys is also mapped to the same shardnet's secret key.

Reply:
Each of the 10 keys will be mapped to a different shardnet secret key.

Like if account `foo.near` has 2 keys `A="ed25519:He7QeRu..."` and `B="ed25519:3KyUucjyG..."`, and say `map_key(A).public()="ed25519:FXXrTXi..."` and `map_key(B).public() = "ed25519:Eo9W44tRM..."` (it doesnt actually, I made up those values), then if the source chain has a `DeleteKey("ed25519:3KyUucjyG...")` tx, we will generate a `DeleteKey("ed25519:Eo9W44tRM...")` tx for the target chain